core/src/main/java/com/datastrato/gravitino/listener/impl/AsyncQueueListener.java

```diff
@@ -0,0 +1,139 @@
+import com.datastrato.gravitino.listener.event.Event;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * AsyncQueueListener acts as event listener, and internally buffer event to a queue, start a
+ * dispatcher thread to dispatch event to the real listeners. For default AsyncQueueListener it may
+ * contain multi listeners share with one queue and dispatcher thread. For other
+ * AsyncQueueDispatchers, contain only one listener.
+ */
+public class AsyncQueueListener implements EventListenerPlugin {
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncQueueListener.class);
+  private static final String NAME_PREFIX = "async-queue-listener-";
+
+  private final List<EventListenerPlugin> eventListeners;
+  private final BlockingQueue<Event> queue;
+  private final Thread asyncProcessor;
+  private final int dispatcherJoinSeconds;
+  private final AtomicBoolean stopped = new AtomicBoolean(false);
+  private final AtomicLong dropEventCounters = new AtomicLong(0);
+  private final AtomicLong lastDropEventCounters = new AtomicLong(0);
+  private Instant lastRecordDropEventTime;
+  private final String asyncQueueListenerName;
+
+  public AsyncQueueListener(
+      List<EventListenerPlugin> listeners,
+      String name,
+      int queueCapacity,
+      int dispatcherJoinSeconds) {
+    this.asyncQueueListenerName = NAME_PREFIX + name;
+    this.eventListeners = listeners;
+    this.queue = new LinkedBlockingQueue<>(queueCapacity);
+    this.asyncProcessor = new Thread(() -> processEvents());
+    this.dispatcherJoinSeconds = dispatcherJoinSeconds;
+    asyncProcessor.setDaemon(true);
+    asyncProcessor.setName(asyncQueueListenerName);
+  }
+
+  @Override
+  public void onPostEvent(Event event) {
+    if (stopped.get()) {
+      LOG.warn(
+          "{} drop event: {}, since AsyncQueueListener is stopped",
+          asyncQueueListenerName,
+          event.getClass().getSimpleName());
+      return;
+    }
+
+    if (queue.offer(event)) {
+      return;
+    }
+
+    long currentDropEvents = dropEventCounters.incrementAndGet();
+    long lastDropEvents = lastDropEventCounters.get();
+    // dropEvents may less than zero in such conditions:
+    // 1. Thread A increment dropEventCounters
+    // 2. Thread B increment dropEventCounters and update lastDropEventCounters
+    // 3. Thread A get lastDropEventCounters
+    long dropEvents = currentDropEvents - lastDropEvents;
+    if (dropEvents > 0 && Instant.now().isAfter(lastRecordDropEventTime.plusSeconds(60))) {
+      if (lastDropEventCounters.compareAndSet(lastDropEvents, currentDropEvents)) {
```

Comment:
It seems that `currentDropEvents` is not equal to the value it is now in some cases.

Reply:
yes,  onPostEvents maybe invoked by multi thread